### PR TITLE
Addressed the credit card number issue

### LINKED ISSUE #110: Storing credit card information in the clear is a problem
Nice work on eshop/Aspire! I was looking through the code to get a better handle on Aspire and other things when I came across this migration:

https://github.com/dotnet-architecture/eShopOnAzure/blob/master/src/Services/Ordering/Ordering.API/Infrastructure/Migrations/20170208181933_Initial.cs#L91

It looks like a table is created in the database with columns to store card number, expiration date, and security code - all in the clear. Looking at the command/model for saving this information:

https://github.com/dotnet/eShop/blob/main/src/Ordering.API/Application/Commands/CreateOrderCommand.cs#L78

It doesn't look like there's any encryption or attempt to conceal private information for the card or the user. Running the app and placing an order confirms this:

screenshot_225

This is kind of a Big Deal when it comes to [PCI Compliance](https://www.nerdwallet.com/article/small-business/pci-compliance) which many companies are held to.  Moreover it's just a good idea to not store credit card information in the clear in a database :). Given that we refer to this as a "reference app" it might be a good idea to _do the needful_ for people so they don't accidentally do Bad Things.

## Recommendations

One idea that we could do here would be to call out, explicitly, in the API that we're not logging/storing the credit card information. At that point you could return a FakePayment object, for instance, with the information the app needs.

Another idea would be to stub out a Stripe Webhook Receiver with Stripe Checkout integration. Easy to do and I would be more than happy to PR that!

### PR DESCRIPTION
**Pull request by robconery**
The credit card number for an order was being stored in the database, saved to the user record as well as payment methods and orders. In addition, the card number was also logged in clear text.

I added a mask at the API request point for a new order, so instead of `4111111111111111` it is now `XXXXXXXXXXXX1111`. This is **not** an optimal solution, but is one step better than storing things in clear text.

Fixes #110

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (timheuer) [inline .vscode/launch.json:1]: I'd remove .vscode launch assets -- the pre-release/may release of the C# Dev Kit has a debugger launcher support for Aspire and children processes.

File: src/Ordering.API/Apis/OrdersApi.cs [hunk 2]
```diff
@@ -135,9 +137,10 @@ public static async Task<OrderDraftDTO> CreateOrderDraftAsync(CreateOrderDraftCo
 
         using (services.Logger.BeginScope(new List<KeyValuePair<string, object>> { new("IdentifiedCommandId", requestId) }))
         {
+            var maskedCCNumber = request.CardNumber.Replace(request.CardNumber.Substring(0, 12), new string('X', 12));
```
Comment (timheuer) [inline src/Ordering.API/Apis/OrdersApi.cs:140]: I have no idea what is more efficient for the compiler, but allocating a new string may not be? I would have used PadLeft but maybe someone tells me that's bad.


`request.CardNumber.Substring(request.CardNumber.Length - 4).PadLeft(request.CardNumber.Length, 'X');`

Comment (robconery) [reply src/Ordering.API/Apis/OrdersApi.cs:140]: Given that it's an order, which isn't a high volume thing, is this really an issue?

File: src/Ordering.API/Apis/OrdersApi.cs [hunk 3]
```diff
@@ -120,12 +120,14 @@ public static async Task<OrderDraftDTO> CreateOrderDraftAsync(CreateOrderDraftCo
...
+        
+        //mask the credit card number
+        
         services.Logger.LogInformation(
             "Sending command: {CommandName} - {IdProperty}: {CommandId} ({@Command})",
             request.GetGenericTypeName(),
             nameof(request.UserId),
-            request.UserId,
-            request);
+            request.UserId, new{}); //don't log the request as it has CC number
```
Comment (ReubenBond) [inline src/Ordering.API/Apis/OrdersApi.cs:130]: Why not remove the `Command` parameter altogether?

Comment (robconery) [reply src/Ordering.API/Apis/OrdersApi.cs:130]: I'm not too sure what you're asking; this code is a bit hard for me to follow and I went with "least impact" wherever I could.

**Comment by ReubenBond** [comment]
Ideally, this would show people how to securely deal with payments and eliminate credit card storage altogether, rather than masking payment details without facilitating implementation of a secure payment flow.

I'm _ok_ with this for now, but it doesn't lead people in the right direction currently.

File: src/Ordering.API/Apis/OrdersApi.cs [hunk 4]
```diff
@@ -120,12 +120,14 @@ public static async Task<OrderDraftDTO> CreateOrderDraftAsync(CreateOrderDraftCo
         CreateOrderRequest request,
         [AsParameters] OrderServices services)
     {
+        
+        //mask the credit card number
+        
         services.Logger.LogInformation(
             "Sending command: {CommandName} - {IdProperty}: {CommandId} ({@Command})",
```
Comment (ReubenBond) [inline src/Ordering.API/Apis/OrdersApi.cs:127]:
```suggestion
            "Sending command: {CommandName} - {IdProperty}: {CommandId}",
```

Comment (ReubenBond) [reply src/Ordering.API/Apis/OrdersApi.cs:127]: This, then delete the `, new{}` on line 130 below

File: src/Ordering.API/Apis/OrdersApi.cs (hunk 3 above)
Comment (ReubenBond) [reply src/Ordering.API/Apis/OrdersApi.cs:130]: I added a diff above to demonstrate

**Comment by robconery** [comment]
I did start down that route (removing CC storage) but I wanted to stay focused on the task at hand, which is _triage_ for Build. Once we get this done, we can move on to a more ideal situation - but the CC bits are in the fabric of the application, meaning that the application requires a user to login, then we attach a payment method to them (which is CC rn), and that payment method ID is attached to the order. This is a weird way of doing eCommerce, but we can think on that when it's not on fire :D

**Comment by ReubenBond** [comment]
This 2-step approach is fine to me. I see that CI is failing, eg the `eShop.Ordering.FunctionalTests.OrderingApiTests.AddNewOrder` test. Could you please fix the tests?

Any help you can provide in getting the payments flow up-to-scratch is much appreciated.

**Comment by robconery** [comment]
Thanks James - should have had a null check in there!

**Comment by adityamandaleeka** [comment]
Looks like this has already been merged, but I'd like to see a comment added here as well that indicates why we're doing the masking and explicitly point out that this isn't the "solution" to this problem. In the interest of guiding people down the right path, we wouldn't want people to come to this code and just remove the masking and move on...

**Pull request merged**